### PR TITLE
fix(security): harden API, WebSocket, validation, and command execution

### PR DESCRIPTION
**Pull request by ruaan-deysel**
## Summary

Comprehensive security hardening based on OWASP Top 10 audit findings. All fixes address real vulnerabilities or defense-in-depth gaps identified during a full codebase security review.

### Security Fixes

| Fix | File(s) | Severity |
|-----|---------|----------|
| **CWE-78: Shell injection in userscripts** — replaced `sh -c` + `fmt.Sprintf` with direct arg passing | `controllers/userscripts.go` | P0 |
| **WebSocket origin validation** — per-request upgrader validates Origin host against request Host | `api/websocket.go` | P1 |
| **WebSocket ReadLimit** — 64 KB limit prevents message-size DoS | `api/websocket.go` | P1 |
| **Security headers** — `X-Content-Type-Options`, `X-Frame-Options`, `CSP`, `Referrer-Policy`, `Permissions-Policy` | `api/middleware.go` | P1 |
| **CSRF origin validation** — validates Origin on POST/PUT/PATCH/DELETE with localhost-aware matching | `api/middleware.go` | P1 |
| **Request body size limit** — 1 MB `MaxBytesReader` on all endpoints | `api/middleware.go` | P2 |
| **HTTP server timeouts** — `ReadHeaderTimeout` (10s), `IdleTimeout` (120s) against slowloris | `api/server.go` | P2 |
| **CWE-158: Null byte injection** — added checks to `ValidateShareName`, `ValidateUserScriptName`, `ValidatePluginName`, `ValidateSnapshotName` | `lib/validation.go` | P2 |
| **Consistent validation errors** — refactored `ValidateLogFilename` from `bool` → `error` return | `lib/validation.go`, `api/handlers.go` | P3 |
| **Direct `exec.Command` elimination** — replaced 3 occurrences with `lib.ExecCommand*` wrappers | `collectors/unassigned.go`, `watchdog/probes.go` | P2 |
| **Race condition fix** — `sync.Mutex` protecting `prevCPU` map in DockerCollector | `collectors/docker.go` | P1 |
| **Fix pre-existing nosec directive** — moved misplaced `#nosec G304` in diagnostics bundle | `diagnostics/bundle.go` | P3 |

### Files Changed (12)

- `daemon/services/controllers/userscripts.go`
- `daemon/services/api/websocket.go`
- `daemon/services/api/middleware.go`
- `daemon/services/api/server.go`
- `daemon/services/api/handlers.go`
- `daemon/lib/validation.go`
- `daemon/lib/validation_test.go`
- `daemon/services/collectors/docker.go`
- `daemon/services/collectors/unassigned.go`
- `daemon/services/watchdog/probes.go`
- `daemon/services/diagnostics/bundle.go`
- `CHANGELOG.md`

## Testing

- [x] All unit tests pass with `-race` detection (15 test suites)
- [x] All 22 pre-commit hooks pass (gofmt, go vet, goimports, golangci-lint, gosec, govulncheck, etc.)
- [x] Full Ansible deploy + verify on Unraid hardware:
  - 58/58 API endpoints pass
  - MCP protocol compliance (init, tools, resources, prompts, call, read)
  - WebSocket upgrade (101 Switching Protocols)
  - Prometheus metrics (45 families)
  - Swagger docs (129 paths)
  - Error handling (8 cases including path traversal)
  - Alerting CRUD + Health Check CRUD
  - 15/15 collectors running

## What's NOT included

- **API key / Bearer token authentication** — intentionally excluded per project scope (local network only, never exposed publicly)

## Summary by CodeRabbit

* **Security**
  * Enhanced HTTP & WebSocket protections: stricter Origin/CSRF checks, security headers, tighter server timeouts
  * Global 1MB request body limit and 64KB WebSocket read limit
  * Null‑byte rejection in name validators and clearer filename validation errors

* **Bug Fixes**
  * Fixed Docker collector race conditions
  * More reliable background-script handling and child-process reaping
  * Added PATCH to CORS allow list; improved ping target validation
  * More robust block-device and external-command handling

### CODE REVIEW

**Comment by ruaan-deysel** [comment]
@coderabbitai review

File: daemon/services/api/middleware.go [hunk 1]
```diff
@@ -73,3 +75,103 @@ func recoveryMiddleware(next http.Handler) http.Handler {
...
+			if err != nil {
+				http.Error(w, "Forbidden: invalid origin", http.StatusForbidden)
+				return
+			}
+
+			if allowedOrigin != "*" {
+				// Strict mode: origin must match configured value
+				if origin == allowedOrigin {
+					next.ServeHTTP(w, r)
+					return
```
[the remark this answers is not in the document]
Comment (ruaan-deysel) [reply daemon/services/api/middleware.go:147]: Fixed in c0bf67d. Added `if allowedOrigin == "" { allowedOrigin = "*" }` at the top of `csrfMiddleware`, consistent with `corsMiddleware` behavior.

File: daemon/services/api/middleware.go [hunk 2]
```diff
@@ -73,3 +75,103 @@ func recoveryMiddleware(next http.Handler) http.Handler {
...
+	}
+}
+
+// stripHostPort removes the port portion from a host:port string.
+func stripHostPort(hostport string) string {
+	host, _, err := net.SplitHostPort(hostport)
+	if err != nil {
+		return hostport // already just a host
+	}
+	return host
```
[the remark this answers is not in the document]
Comment (ruaan-deysel) [reply daemon/services/api/middleware.go]: Fixed in c0bf67d. Added bracketed IPv6 normalization: when `net.SplitHostPort` fails and the input is `[addr]`, we now strip the brackets so it matches `url.URL.Hostname()` output.

File: daemon/services/api/middleware.go [hunk 3]
```diff
@@ -73,3 +75,103 @@ func recoveryMiddleware(next http.Handler) http.Handler {
...
+				// Non-browser clients (curl, API tools) don't send Origin
+				next.ServeHTTP(w, r)
+				return
+			}
+
+			// Parse and validate origin
+			parsed, err := url.Parse(origin)
+			if err != nil {
+				http.Error(w, "Forbidden: invalid origin", http.StatusForbidden)
+				return
```
[the remark this answers is not in the document]
Comment (ruaan-deysel) [reply daemon/services/api/middleware.go:140]: Fixed in c0bf67d. All three `http.Error` calls in CSRF middleware now use `respondJSON` with a JSON error map for consistent API responses.

File: daemon/services/api/websocket.go [hunk 4]
```diff
@@ -191,12 +189,33 @@ func (h *WSHub) Broadcast(topic string, data any) {
...
+		CheckOrigin: func(req *http.Request) bool {
+			origin := req.Header.Get("Origin")
+			if origin == "" {
+				return true // non-browser clients don't send Origin
+			}
+			if s.ctx.CORSOrigin == "*" {
+				// In wildcard mode, verify origin host matches request host
+				// to prevent drive-by CSRF from arbitrary external sites.
+				parsed, err := url.Parse(origin)
+				if err != nil {
```
[the remark this answers is not in the document]
Comment (ruaan-deysel) [reply daemon/services/api/websocket.go]: Fixed in c0bf67d. WebSocket `CheckOrigin` now normalizes empty `CORSOrigin` to `"*"` before the wildcard check, matching `corsMiddleware` and `csrfMiddleware` behavior.

File: daemon/services/controllers/userscripts.go [hunk 5]
```diff
@@ -110,9 +110,9 @@ func ExecuteUserScript(scriptName string, background bool, wait bool) (*dto.User
...
 // executeScriptBackground executes a script in the background
 func executeScriptBackground(scriptPath string, scriptName string) (*dto.UserScriptExecuteResponse, error) {
-	// Execute script in background using bash with nohup to detach
-	// We use sh -c to run the command in background
-	_, err := lib.ExecCommand("sh", "-c", fmt.Sprintf("nohup bash %s > /dev/null 2>&1 &", scriptPath))
+	// Execute script in background using nohup + bash with direct arguments
+	// to avoid shell interpolation of the script path (CWE-78 prevention)
+	_, err := lib.ExecCommand("nohup", "bash", scriptPath)
 
 	if err != nil {
```
[the remark this answers is not in the document]
Comment (ruaan-deysel) [reply daemon/services/controllers/userscripts.go]: Fixed in c0bf67d. Replaced `lib.ExecCommand` with `os.StartProcess` + `/dev/null` stdio + `proc.Release()` to truly background the script without blocking the API handler or risking zombie processes.

File: daemon/services/api/middleware.go [hunk 6]
```diff
@@ -73,3 +75,103 @@ func recoveryMiddleware(next http.Handler) http.Handler {
...
 	})
 }
+
+// securityHeadersMiddleware adds standard security headers to every response.
+func securityHeadersMiddleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Set("X-Content-Type-Options", "nosniff")
+		w.Header().Set("X-Frame-Options", "DENY")
+		w.Header().Set("X-XSS-Protection", "1; mode=block")
+		w.Header().Set("Content-Security-Policy", "default-src 'self'")
```
[the remark this answers is not in the document]
Comment (ruaan-deysel) [reply daemon/services/api/middleware.go]: Fixed in c0bf67d. Added path-based CSP: `/swagger/` routes now get a relaxed policy permitting `'unsafe-inline'` for scripts/styles and `data:` for images, while all other routes keep `default-src 'self'`.

File: CHANGELOG.md [hunk 7]
```diff
@@ -9,6 +9,34 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Security
+
+- **Fix shell injection in userscripts controller** — replaced `sh -c` with `fmt.Sprintf` pattern
+  with direct argument passing to eliminate CWE-78 command injection risk
```
[the remark this answers is not in the document]
Comment (ruaan-deysel) [reply CHANGELOG.md]: Fixed in c0bf67d. Reworded to: "removed the `sh -c` + `fmt.Sprintf` command-construction pattern in favor of direct argument passing."

File: daemon/services/controllers/userscripts.go [hunk 8]
```diff
@@ -108,20 +108,45 @@ func ExecuteUserScript(scriptName string, background bool, wait bool) (*dto.User
...
 		return &dto.UserScriptExecuteResponse{
 			Success: false,
 			Error:   fmt.Sprintf("Failed to execute script: %v", err),
 		}, err
 	}
 
+	// Release the process so it doesn't become a zombie
+	if err := proc.Release(); err != nil {
+		logger.Warning("Failed to release background process for script %s: %v", scriptName, err)
+	}
```
[the remark this answers is not in the document]
Comment (ruaan-deysel) [reply daemon/services/controllers/userscripts.go]: Fixed in cbc014e. Replaced `proc.Release()` with `go func() { _, _ = proc.Wait() }()` to properly reap the child process and prevent zombies.

File: daemon/services/collectors/unassigned.go [hunk 9]
```diff
@@ -176,14 +176,13 @@ func (c *UnassignedCollector) getArrayDisks() map[string]bool {
...
-	cmd := exec.Command("lsblk", "-d", "-n", "-o", "NAME")
-	output, err := cmd.Output()
+	output, err := lib.ExecCommandOutput("lsblk", "-d", "-n", "-o", "NAME")
 	if err != nil {
 		logger.Error("Failed to list block devices: %v", err)
 		return []string{}
 	}
 
-	lines := strings.Split(strings.TrimSpace(string(output)), "\n")
+	lines := strings.Split(strings.TrimSpace(output), "\n")
```
[the remark this answers is not in the document]
Comment (ruaan-deysel) [reply daemon/services/collectors/unassigned.go:185]: Fixed in cbc014e. Added `lib.ExecCommandStdout()` helper that uses `cmd.Output()` (stdout-only) instead of `CombinedOutput()`. Switched `getDeviceInfo` lsblk JSON call to use it so stderr warnings don't corrupt parsing.

File: daemon/services/api/server.go [hunk 10]
```diff
@@ -88,9 +88,12 @@ func NewServerWithCollectorManager(ctx *domain.Context, cm CollectorManagerInter
 
 func (s *Server) setupRoutes() {
 	// Apply middleware
+	s.router.Use(recoveryMiddleware)
+	s.router.Use(securityHeadersMiddleware)
 	s.router.Use(corsMiddleware(s.ctx.CORSOrigin))
+	s.router.Use(csrfMiddleware(s.ctx.CORSOrigin))
+	s.router.Use(bodySizeLimitMiddleware)
```
[the remark this answers is not in the document]
Comment (ruaan-deysel) [reply daemon/services/api/server.go:95]: Fixed in cbc014e. Added `PATCH` to `Access-Control-Allow-Methods` in `corsMiddleware`, and updated the corresponding test expectation.

File: daemon/services/api/websocket.go [hunk 11]
```diff
@@ -191,12 +189,38 @@ func (h *WSHub) Broadcast(topic string, data any) {
...
+					return false
+				}
+				return parsed.Hostname() == stripHostPort(req.Host)
+			}
+			return origin == corsOrigin
+		},
+	}
```
[the remark this answers is not in the document]
Comment (ruaan-deysel) [reply daemon/services/api/websocket.go:260]: Fixed in cbc014e. WebSocket `CheckOrigin` now uses `isLocalhost()` equivalence check matching `csrfMiddleware` behavior — `localhost`, `127.0.0.1`, and `::1` are treated as equivalent.

File: daemon/services/watchdog/probes.go [hunk 12]
```diff
@@ -79,10 +79,9 @@ func probeTCP(_ context.Context, target string, timeout time.Duration) ProbeResu
 // Target should be a hostname or IP address.
 func probePing(ctx context.Context, target string, timeout time.Duration) ProbeResult {
 	timeoutSec := fmt.Sprintf("%d", max(1, int(timeout.Seconds())))
-	// #nosec G204 -- Target is passed as a direct argv value to ping without shell interpolation.
-	cmd := exec.CommandContext(ctx, "ping", "-c", "1", "-W", timeoutSec, target)
-	if output, err := cmd.CombinedOutput(); err != nil {
-		return ProbeResult{Healthy: false, Error: fmt.Sprintf("ping failed: %s (%s)", err, string(output))}
+	output, err := lib.ExecCommandOutputWithContext(ctx, "ping", "-c", "1", "-W", timeoutSec, target)
+	if err != nil {
+		return ProbeResult{Healthy: false, Error: fmt.Sprintf("ping failed: %s (%s)", err, output)}
```
[the remark this answers is not in the document]
Comment (ruaan-deysel) [reply daemon/services/watchdog/probes.go:87]: Fixed in cbc014e. Added `lib.ValidateHostOrIP()` that rejects empty strings, hyphen-prefixed values (flag injection), and anything that isn't a valid IP or RFC-1123 hostname. Applied in `probePing` before executing the ping command.

**Comment by ruaan-deysel** [comment]
@copilot apply changes based on the comments in [this thread](https://github.com/ruaan-deysel/unraid-management-agent/pull/106#pullrequestreview-4086253573)

File: daemon/lib/shell.go [hunk 13]
```diff
@@ -77,6 +77,21 @@ func ExecCommandOutputWithContext(ctx context.Context, command string, args ...s
 	return string(output), nil
 }
 
+// ExecCommandStdout executes a command and returns only stdout, discarding
+// stderr. Use this instead of ExecCommandOutput when the output is
+// machine-parsed (e.g. JSON) and stderr warnings could contaminate results.
+func ExecCommandStdout(command string, args ...string) (string, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
```
[the remark this answers is not in the document]
Comment (ruaan-deysel) [reply daemon/lib/shell.go:88]: Tests added in commit 01bba42. `TestExecCommandStdout` in `shell_test.go` covers: stdout returned on success, stderr is discarded (uses `bash -c` to emit both stdout and stderr, verifies only stdout appears), non-zero exit code returns wrapped error, and non-existent command returns error.

File: daemon/services/api/middleware.go [hunk 14]
```diff
@@ -73,3 +75,156 @@ func recoveryMiddleware(next http.Handler) http.Handler {
...
+}
+
+// isLocalhost returns true if the host is a loopback address or "localhost".
+func isLocalhost(host string) bool {
+	switch strings.ToLower(host) {
+	case "localhost", "127.0.0.1", "::1":
+		return true
+	}
+	return false
+}
```
[the remark this answers is not in the document]
Comment (ruaan-deysel) [reply daemon/services/api/middleware.go]: Fixed in db45cc0. Added missing trailing newline to `middleware.go` — `gofmt` now reports no diff.

File: daemon/services/api/websocket.go [hunk 15]
```diff
@@ -284,4 +354,4 @@ func (c *WSClient) readPump() {
 		c.setTopics(topics) // nil means "all topics"
 		logger.Debug("WebSocket client updated topic filter: %v", topics)
 	}
-}
+}
```
[the remark this answers is not in the document]
Comment (ruaan-deysel) [reply daemon/services/api/websocket.go]: Fixed in db45cc0. Added missing trailing newline to `websocket.go` — `gofmt` now reports no diff.

File: daemon/lib/validation.go [hunk 16]
```diff
@@ -375,3 +396,29 @@ func ValidateCPUGovernor(governor string) error {
...
+
+// hostnameRegex allows RFC-952/1123 hostnames: labels of alphanumerics and hyphens
+// separated by dots, each label at most 63 chars, total at most 253 chars.
+var hostnameRegex = regexp.MustCompile(`^(?i)[a-z0-9]([a-z0-9-]{0,61}[a-z0-9])?(\.[a-z0-9]([a-z0-9-]{0,61}[a-z0-9])?)*$`)
+
+// ValidateHostOrIP validates that a string is a valid hostname or IP address.
+// It rejects empty strings, strings starting with '-' (flag injection), and
+// values that are not valid IPs or RFC-1123 hostnames.
+func ValidateHostOrIP(host string) error {
+	if host == "" {
```
[the remark this answers is not in the document]
Comment (ruaan-deysel) [reply daemon/lib/validation.go:408]: Tests added in commit 01bba42. `TestValidateHostOrIP` in `validation_test.go` covers: valid hostnames, IPv4/IPv6, empty input, leading-hyphen flag injection, whitespace, shell metacharacters, null bytes, excessive length, and edge cases (trailing dots, label-ending hyphens).

File: daemon/services/api/websocket.go [hunk 17]
```diff
@@ -191,12 +189,44 @@ func (h *WSHub) Broadcast(topic string, data any) {
 //	@Tags			WebSocket
 //	@Router			/ws [get]
 func (s *Server) handleWebSocket(w http.ResponseWriter, r *http.Request) {
-	conn, err := upgrader.Upgrade(w, r, nil)
+	u := websocket.Upgrader{
+		CheckOrigin: func(req *http.Request) bool {
+			origin := req.Header.Get("Origin")
+			if origin == "" {
```
[the remark this answers is not in the document]
Comment (ruaan-deysel) [reply daemon/services/api/websocket.go:196]: Tests added in commit 01bba42. `websocket_coverage_test.go` now includes: `TestWebSocketOriginWildcardSameHost`, `TestWebSocketOriginWildcardMismatch`, `TestWebSocketOriginNoHeader`, `TestWebSocketOriginSpecificMatch`, `TestWebSocketOriginSpecificMismatch` — covering same-host acceptance, mismatched origin rejection (403), no-Origin passthrough, specific origin match, and specific origin mismatch.

File: daemon/lib/validation.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (ruaan-deysel) [reply daemon/lib/validation.go:408]: Tests added in commit 01bba42. `TestValidateHostOrIP` in `validation_test.go` covers: valid hostnames, IPv4/IPv6, empty input, leading-hyphen flag injection, whitespace, shell metacharacters, null bytes, excessive length, and edge cases (trailing dots, label-ending hyphens).

File: daemon/lib/shell.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (ruaan-deysel) [reply daemon/lib/shell.go:88]: Tests added in commit 01bba42. `TestExecCommandStdout` in `shell_test.go` covers: stdout returned on success, stderr is discarded (uses `bash -c` to emit both stdout and stderr, verifies only stdout appears), non-zero exit code returns wrapped error, and non-existent command returns error.

File: daemon/services/api/websocket.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (ruaan-deysel) [reply daemon/services/api/websocket.go:196]: Tests added in commit 01bba42. `websocket_coverage_test.go` now includes: `TestWebSocketOriginWildcardSameHost`, `TestWebSocketOriginWildcardMismatch`, `TestWebSocketOriginNoHeader`, `TestWebSocketOriginSpecificMatch`, `TestWebSocketOriginSpecificMismatch` — covering same-host acceptance, mismatched origin rejection (403), no-Origin passthrough, specific origin match, and specific origin mismatch.

File: daemon/services/api/websocket.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (ruaan-deysel) [reply daemon/services/api/websocket.go]: Fixed in db45cc0. Added missing trailing newline to `websocket.go` — `gofmt` now reports no diff.

File: daemon/services/api/middleware.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (ruaan-deysel) [reply daemon/services/api/middleware.go]: Fixed in db45cc0. Added missing trailing newline to `middleware.go` — `gofmt` now reports no diff.
